crates/turbo-tasks-memory/src/aggregation/balance_edge.rs

```diff
@@ -0,0 +1,206 @@
+    // The last info about followers
+    let mut followers_count = None;
+
+    loop {
+        let root = upper_aggregation_number == u32::MAX || target_aggregation_number == u32::MAX;
+        let order = if root {
+            Ordering::Greater
+        } else {
+            upper_aggregation_number.cmp(&target_aggregation_number)
+        };
+        match order {
+            Ordering::Equal => {
+                // we probably want to increase the aggregation number of target
+                let upper = ctx.node(upper_id);
+                upper_aggregation_number = upper.aggregation_number();
+                drop(upper);
+                if upper_aggregation_number != u32::MAX
+                    && upper_aggregation_number == target_aggregation_number
+                {
+                    let target = ctx.node(target_id);
+                    target_aggregation_number = target.aggregation_number();
+                    if upper_aggregation_number == target_aggregation_number {
+                        // increase target aggregation number
+                        increase_aggregation_number_internal(
+                            ctx,
+                            balance_queue,
+                            target,
+                            target_id,
+                            target_aggregation_number + 1,
+                            target_aggregation_number + 1,
+                        );
+                    }
+                }
+            }
+            Ordering::Less => {
+                // target should probably be a follower of upper
+                if uppers_count.map_or(false, |count| count <= 0) {
+                    // We already removed all uppers, maybe too many
+                    break;
+                } else if extra_followers == 0 {
+                    let upper = ctx.node(upper_id);
+                    upper_aggregation_number = upper.aggregation_number();
+                    if upper_aggregation_number < target_aggregation_number {
+                        // target should be a follower of upper
+                        // add some extra followers
+                        let count = uppers_count.unwrap_or(1) as usize;
+                        extra_followers += count;
+                        followers_count = Some(add_follower_count(
+                            ctx,
+                            balance_queue,
+                            upper,
+                            upper_id,
+                            target_id,
+                            count,
+                            true,
+                        ));
+                    }
+                } else {
+                    // we already have extra followers, remove some uppers to balance
+                    let count = extra_followers + extra_uppers;
+                    let target = ctx.node(target_id);
+                    if is_in_progress(ctx, upper_id) {
+                        drop(target);
```

Comment:
What is the significance of borrowing before the branch and then dropping?

Reply:
We need to check in progress (which is an atomic) while either holding the target or the upper lock. In progress is only set during the `target` lock, so when we read it it need to be under the `target` lock. If not in progress we can continue working with the target in the else branch.

Otherwise we want to enqueue our work to the upper node. So we acquire a upper lock. In the meantime the in_progress flag might have changed, so we need to check that again.